src/herder/HerderImpl.cpp

```diff
@@ -1024,6 +1028,67 @@ HerderImpl::getCurrentlyTrackedQuorum() const
     return mPendingEnvelopes.getCurrentlyTrackedQuorum();
 }
 
+static Hash
+getQmapHash(QuorumTracker::QuorumMap const& qmap)
+{
+    std::unique_ptr<SHA256> hasher = SHA256::create();
+    for (auto const& pair : qmap)
+    {
+        hasher->add(xdr::xdr_to_opaque(pair.first));
+        if (pair.second)
+        {
+            hasher->add(xdr::xdr_to_opaque(*pair.second));
+        }
+        else
+        {
+            hasher->add("\0");
+        }
+    }
+    return hasher->finish();
+}
+
+void
+HerderImpl::checkAndMaybeReanalyzeQuorumMap()
+{
+    using CS = QuorumMapIntersectionState::CheckingState;
+
+    if (mLastQuorumMapIntersectionState.mCheckingState != CS::RUNNING_CHECK)
+    {
+        QuorumTracker::QuorumMap const& qmap = getCurrentlyTrackedQuorum();
+        Hash curr = getQmapHash(qmap);
+        if (mLastQuorumMapIntersectionState.mQuorumMapHash != curr)
+        {
+            CLOG(INFO, "Herder")
+                << "Transitive closure of quorum has changed, re-analyzing.";
+            mLastQuorumMapIntersectionState.mCheckingState = CS::RUNNING_CHECK;
+            mLastQuorumMapIntersectionState.mLedgerSeq = getCurrentLedgerSeq();
+            mLastQuorumMapIntersectionState.mQuorumMapHash = curr;
+            auto& cfg = mApp.getConfig();
+            auto qic = QuorumIntersectionChecker::create(qmap, cfg);
+            auto& hState = mLastQuorumMapIntersectionState;
+            auto& app = mApp;
+            auto worker = [qic, &app, &hState] {
+                bool ok = qic->networkEnjoysQuorumIntersection();
+                app.postOnMainThread(
+                    [ok, &hState] {
+                        if (ok)
```

Comment:
because it's writing to a variable (hState) shared with the main thread. posting the write back to the main thread serializes access. otherwise we're racing.

Reply:
ah yes good point, didn't think that it was a shared structure